pkg/controller/memberstatus/memberstatus_controller.go

```diff
@@ -0,0 +1,236 @@
+			// MemberStatus resource not found, log error and return
+			reqLogger.Error(err, "MemberStatus resource was not found")
+		}
+		return reconcile.Result{}, nil
+	}
+
+	err = r.aggregateAndUpdateStatus(reqLogger, memberStatus)
+	if err != nil {
+		reqLogger.Error(err, "Failed to update status")
+	}
+
+	reqLogger.Info(fmt.Sprintf("Finished updating MemberStatus, requeueing after %v", defaultRequeueTime))
+	return reconcile.Result{RequeueAfter: defaultRequeueTime}, nil
+}
+
+type statusHandler struct {
+	Name         statusComponentTag
+	handleStatus func(logger logr.Logger, memberStatus *toolchainv1alpha1.MemberStatus) error
+}
+
+func (r *ReconcileMemberStatus) aggregateAndUpdateStatus(reqLogger logr.Logger, memberStatus *toolchainv1alpha1.MemberStatus) error {
+
+	memberOperatorStatusHandler := statusHandler{Name: memberOperator, handleStatus: r.memberOperatorHandleStatus}
+	hostConnectionStatusHandler := statusHandler{Name: hostConnection, handleStatus: r.hostConnectionHandleStatus}
+
+	statusHandlers := []statusHandler{memberOperatorStatusHandler, hostConnectionStatusHandler}
+
+	// track components that are not ready
+	unreadyComponents := []string{}
+
+	// retrieve component statuses eg. kubefed, member deployment
+	for _, handler := range statusHandlers {
+		err := handler.handleStatus(reqLogger, memberStatus)
+		if err != nil {
+			reqLogger.Error(err, "status update problem")
+			unreadyComponents = append(unreadyComponents, string(handler.Name))
+		}
+	}
+
+	// if any components were not ready then set the overall status to not ready
+	if len(unreadyComponents) > 0 {
+		err := fmt.Errorf("Components not ready: %v", unreadyComponents)
+		return r.setStatusNotReady(memberStatus, err.Error())
+	}
+	return r.setStatusReady(memberStatus)
+}
+
+func (r *ReconcileMemberStatus) hostConnectionHandleStatus(reqLogger logr.Logger, memberStatus *toolchainv1alpha1.MemberStatus) error {
+	// look up host connection status
+	fedCluster, ok := r.getHostCluster()
+	if !ok {
+		return fmt.Errorf("the host connection was not found")
+	}
+	memberStatus.Status.HostConnection = *fedCluster.ClusterStatus.DeepCopy()
+
+	// check conditions of host connection
+	if !util.IsClusterReady(fedCluster.ClusterStatus) {
+		return fmt.Errorf("the host connection is not ready")
+	}
+
+	return nil
+}
+
+func (r *ReconcileMemberStatus) memberOperatorHandleStatus(reqLogger logr.Logger, memberStatus *toolchainv1alpha1.MemberStatus) error {
+	operatorStatus := toolchainv1alpha1.MemberOperatorStatus{
+		Version:        version.Version,
+		Revision:       version.Commit,
+		BuildTimestamp: version.BuildTime,
+	}
+
+	// look up status of member deployment
+	memberDeploymentName := types.NamespacedName{Namespace: memberStatus.Namespace, Name: memberOperatorDeploymentName}
+	memberDeployment := &appsv1.Deployment{}
+	err := r.client.Get(context.TODO(), memberDeploymentName, memberDeployment)
+	if err != nil {
+		return fmt.Errorf("the member operator deployment was not found")
```

Comment:
It can be NotFound. Or it can be something else (a client error). Do not hide the original error.

Reply:
Good catch! I wrapped the error now: f617c6207d45e7f528f250ccae0ac2fe72fa17d6